### ISSUE TITLE
Segmentation violation while converting a TTree to RNTuple

### ISSUE THREAD

**Issue by ktf** [root-project/root#13996]
### Check duplicate issues.

- [ ] Checked for duplicates

### Description

TL;DR: I did a self contained executable which you can find in the Reproducer part and it crashes using  <http://alimonitor.cern.ch/train-workdir/testdata/LFN/alice/data/2015/LHC15o/000246392/pass2/PWGZZ/Run3_Conversion/320_20220701-0813/AOD/174/AO2D.root> with the following stacktrace:

```
TDirectoryFile
Found TDirectoryFile: DF_2463920048222775916
Found TTree: DbgEventExtra
Convert leaf fStart [Int_t] --> field fStart [std::int32_t]
Convert leaf fNentries [Int_t] --> field fNentries [std::int32_t]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Found TTree: O2bc
Convert leaf fRunNumber [Int_t] --> field fRunNumber [std::int32_t]
Convert leaf fGlobalBC [ULong64_t] --> field fGlobalBC [std::uint64_t]
Convert leaf fTriggerMask [ULong64_t] --> field fTriggerMask [std::uint64_t]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Found TTree: O2calo
Convert leaf fIndexBCs [Int_t] --> field fIndexBCs [std::int32_t]
Convert leaf fCellNumber [Short_t] --> field fCellNumber [std::int16_t]
Convert leaf fAmplitude [Float_t] --> field fAmplitude [float]
Convert leaf fTime [Float_t] --> field fTime [float]
Convert leaf fCellType [Char_t] --> field fCellType [char]
Convert leaf fCaloType [Char_t] --> field fCaloType [char]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Wrote 100000 entries
Wrote 200000 entries
Wrote 300000 entries
Wrote 400000 entries
Found TTree: O2calotrigger
Convert leaf fIndexBCs [Int_t] --> field fIndexBCs [std::int32_t]
Convert leaf fFastOrAbsID [Short_t] --> field fFastOrAbsID [std::int16_t]
Convert leaf fLnAmplitude [Short_t] --> field fLnAmplitude [std::int16_t]
Convert leaf fTriggerBits [Int_t] --> field fTriggerBits [std::int32_t]
Convert leaf fCaloType [Char_t] --> field fCaloType [char]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Wrote 100000 entries
Wrote 200000 entries
Wrote 300000 entries
Found TTree: O2cascade_001
Convert leaf fIndexCollisions [Int_t] --> field fIndexCollisions [std::int32_t]
Convert leaf fIndexV0s [Int_t] --> field fIndexV0s [std::int32_t]
Convert leaf fIndexTracks [Int_t] --> field fIndexTracks [std::int32_t]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Found TTree: O2collision
Convert leaf fIndexBCs [Int_t] --> field fIndexBCs [std::int32_t]
Convert leaf fPosX [Float_t] --> field fPosX [float]
Convert leaf fPosY [Float_t] --> field fPosY [float]
Convert leaf fPosZ [Float_t] --> field fPosZ [float]
Convert leaf fCovXX [Float_t] --> field fCovXX [float]
Convert leaf fCovXY [Float_t] --> field fCovXY [float]
Convert leaf fCovXZ [Float_t] --> field fCovXZ [float]
Convert leaf fCovYY [Float_t] --> field fCovYY [float]
Convert leaf fCovYZ [Float_t] --> field fCovYZ [float]
Convert leaf fCovZZ [Float_t] --> field fCovZZ [float]
Convert leaf fFlags [UShort_t] --> field fFlags [std::uint16_t]
Convert leaf fChi2 [Float_t] --> field fChi2 [float]
Convert leaf fNumContrib [UShort_t] --> field fNumContrib [std::uint16_t]
Convert leaf fCollisionTime [Float_t] --> field fCollisionTime [float]
Convert leaf fCollisionTimeRes [Float_t] --> field fCollisionTimeRes [float]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Found TTree: O2fdd_001
Convert leaf fIndexBCs [Int_t] --> field fIndexBCs [std::int32_t]
Convert leaf fChargeA [Int_t] --> field fChargeA [std::int32_t]
Convert leaf fChargeC [Int_t] --> field fChargeC [std::int32_t]
Convert leaf fTimeA [Float_t] --> field fTimeA [float]
Convert leaf fTimeC [Float_t] --> field fTimeC [float]
Convert leaf fTriggerMask [UChar_t] --> field fTriggerMask [std::uint8_t]
Warning in <[ROOT.NTuple] Warning /Users/ktf/src/sw/SOURCES/ROOT/186ec88ba4/0/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const RNTupleWriteOptions &)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
=================================================================
==5670==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x00010b2300b0 in thread T0
    #0 0x106a86ce0 in wrap_free+0x98 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x52ce0)
    #1 0x1062d3ed8 in ROOT::Experimental::REntry::~REntry() REntry.cxx:25
    #2 0x1046a7cf4 in std::__1::default_delete<ROOT::Experimental::REntry>::operator()[abi:v160006](ROOT::Experimental::REntry*) const unique_ptr.h:65
    #3 0x1046a7c68 in std::__1::unique_ptr<ROOT::Experimental::REntry, std::__1::default_delete<ROOT::Experimental::REntry>>::reset[abi:v160006](ROOT::Experimental::REntry*) unique_ptr.h:297
    #4 0x1046a7b7c in std::__1::unique_ptr<ROOT::Experimental::REntry, std::__1::default_delete<ROOT::Experimental::REntry>>::~unique_ptr[abi:v160006]() unique_ptr.h:263
    #5 0x10469cb0c in std::__1::unique_ptr<ROOT::Experimental::REntry, std::__1::default_delete<ROOT::Experimental::REntry>>::~unique_ptr[abi:v160006]() unique_ptr.h:263
    #6 0x10469a73c in main ao2dToAO2NT.cxx:163
    #7 0x186307f24  (<unknown module>)

0x00010b2300b0 is located 28 bytes after 4-byte region [0x00010b230090,0x00010b230094)
freed by thread T0 here:
    #0 0x106a86ce0 in wrap_free+0x98 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x52ce0)
    #1 0x1062d3ed8 in ROOT::Experimental::REntry::~REntry() REntry.cxx:25
    #2 0x1046a7cf4 in std::__1::default_delete<ROOT::Experimental::REntry>::operator()[abi:v160006](ROOT::Experimental::REntry*) const unique_ptr.h:65
    #3 0x1046a7c68 in std::__1::unique_ptr<ROOT::Experimental::REntry, std::__1::default_delete<ROOT::Experimental::REntry>>::reset[abi:v160006](ROOT::Experimental::REntry*) unique_ptr.h:297
    #4 0x1046a7b7c in std::__1::unique_ptr<ROOT::Experimental::REntry, std::__1::default_delete<ROOT::Experimental::REntry>>::~unique_ptr[abi:v160006]() unique_ptr.h:263
    #5 0x10469cb0c in std::__1::unique_ptr<ROOT::Experimental::REntry, std::__1::default_delete<ROOT::Experimental::REntry>>::~unique_ptr[abi:v160006]() unique_ptr.h:263
    #6 0x10469a73c in main ao2dToAO2NT.cxx:163
    #7 0x186307f24  (<unknown module>)

previously allocated by thread T0 here:
    #0 0x106a86ba4 in wrap_malloc+0x94 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x52ba4)
    #1 0x1062c0d60 in ROOT::Experimental::Detail::RFieldBase::GenerateValue() RField.cxx:332
    #2 0x1062efd48 in ROOT::Experimental::RNTupleModel::CreateEntry() const RNTupleModel.cxx:147
    #3 0x104699f00 in main ao2dToAO2NT.cxx:145
    #4 0x186307f24  (<unknown module>)

SUMMARY: AddressSanitizer: bad-free (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x52ce0) in wrap_free+0x98
==5670==ABORTING
[1]    5670 abort      ./a.out AO2D.root
```

Besides the crash, i would appreciate if one could show me how to create RNTuples in a given folder (which is probably something trivial I am missing).

### Reproducer

```C++
#include <iostream>
#include <getopt.h>
#include <TFile.h>
#include <TTree.h>
#include <TKey.h>
#include <TLeaf.h>
#include <ROOT/RNTuple.hxx>
#include <ROOT/RNTupleModel.hxx>
#include <ROOT/RField.hxx>

void print_help()
{
  std::cout << "Usage: ao2dToAO2NT [options] input_file" << std::endl;
  std::cout << "Options:" << std::endl;
  std::cout << "  -o, --output-file <file>  Output file name" << std::endl;
  std::cout << "  -h, --help                Print this help message" << std::endl;
  std::cout << "  -v, --verbose             Verbose output" << std::endl;
}

bool verbose = false;
std::string outputFile = "ntuple.root";
std::string inputFile;

void debug(const std::string& message)
{
  if (verbose) {
    std::cout << message << std::endl;
  }
}

void log(const std::string& message)
{
  std::cout << message << std::endl;
}

void error(const std::string& message)
{
  std::cerr << message << std::endl;
  exit(1);
}

int main(int argc, char* argv[])
{
  // Use getopt_long to parse the command line
  // the first positional argument is the input file.
  // -o is the output file
  // -h is the help message
  // -v is the verbose flag
  static struct option long_options[] = {
    {"output-file", required_argument, nullptr, 'o'},
    {"help", no_argument, nullptr, 'h'},
    {"verbose", no_argument, nullptr, 'v'},
    {nullptr, 0, nullptr, 0}};
  int option_index = 0;
  while (true) {
    int c = getopt_long(argc, argv, "o:hv", long_options, &option_index);
    // If the help flag is set, print the help message and exit
    if (c == 'h') {
      print_help();
      return 0;
    }
    // If the verbose flag is set, print the verbose message
    else if (c == 'v') {
      verbose = true;
    }
    // If the output file is set, print the output file name
    else if (c == 'o') {
      outputFile = optarg;

    } else if (c == -1) {
      break;
    } else if (c == '?') {
      print_help();
      return 1;
    } else {
      print_help();
      return 1;
    }
  }
  if (optind < argc) {
    inputFile = argv[optind];
  }
  debug("Output file: " + outputFile);
  if (inputFile.empty()) {
    error("Missing input file");
  }
  debug("Input file: " + inputFile);

  // Open the input file using ROOT, iterate over all the TFolders starting with DF_ and
  // convert the TTrees inside such folder an RNTuple in the output file in the same folder.

  TFile* in = TFile::Open(inputFile.c_str(), "READ");
  if (!in) {
    error("Cannot open input file");
  }
  TFile* out = TFile::Open(outputFile.c_str(), "RECREATE");
  if (!out) {
    error("Cannot open output file");
  }
  // Iterate over the directories in the file
  TIter next(in->GetListOfKeys());
  TKey* key;
  while ((key = (TKey*)next())) {
    std::cout << key->GetClassName() << std::endl;
    if (std::string(key->GetClassName()) == "TDirectoryFile") {
      std::string name = key->GetName();
      log("Found TDirectoryFile: " + name);
      auto* dir = (TDirectoryFile*)key->ReadObj();
      if (!out->GetDirectory(name.c_str())) {
        out->mkdir(name.c_str());
      }
      out->cd(name.c_str());
      // Iterate over the objects in the directory
      TIter next2(dir->GetListOfKeys());
      TKey* key2;
      while ((key2 = (TKey*)next2())) {
        if (std::string(key2->GetClassName()) == "TTree") {
          std::string name2 = key2->GetName();
          log("Found TTree: " + name2);
          auto* tree = (TTree*)key2->ReadObj();
          debug("Converting to RNTuple");
          // Create the folder in the output file if it does not exist
          using namespace ROOT::Experimental;
          // Create a model from the input tree
          auto model = RNTupleModel::CreateBare();

          for (auto b : TRangeDynCast<TBranch>(*tree->GetListOfBranches())) {
            // The dynamic cast to TBranch should never fail for GetListOfBranches()
            assert(b);

            // We assume every branch has a single leaf
            auto* l = static_cast<TLeaf*>(b->GetListOfLeaves()->First());

            // Create an ntuple field with the same name and type than the tree branch
            auto field = Detail::RFieldBase::Create(l->GetName(), l->GetTypeName()).Unwrap();
            std::cout << "Convert leaf " << l->GetName() << " [" << l->GetTypeName() << "]"
                      << " --> "
                      << "field " << field->GetName() << " [" << field->GetType() << "]" << std::endl;

            // Hand over ownership of the field to the ntuple model.  This will also create a memory location attached
            // to the model's default entry, that will be used to place the data supposed to be written
            model->AddField(std::move(field));
          }
          auto ntuple = RNTupleWriter::Append(std::move(model), name2, *out);
          auto entry = ntuple->GetModel()->CreateEntry();

          for (auto b : TRangeDynCast<TBranch>(*tree->GetListOfBranches())) {
            auto l = static_cast<TLeaf*>(b->GetListOfLeaves()->First());
            // We connect the model's default entry's memory location for the new field to the branch, so that we can
            // fill the ntuple with the data read from the TTree
            void* fieldDataPtr = entry->GetValue(l->GetName()).GetRawPtr();
            tree->SetBranchAddress(b->GetName(), fieldDataPtr);
          }
          auto nEntries = tree->GetEntries();
          for (decltype(nEntries) i = 0; i < nEntries; ++i) {
            tree->GetEntry(i);
            ntuple->Fill(*entry);

            if (i && i % 100000 == 0) {
              std::cout << "Wrote " << i << " entries" << std::endl;
            }
          }
        }
      }
      out->cd();
    }
  }
}
```

you can then compile it with 

```
c++ -std=c++17 -g ao2dToAO2NT.cxx -I$ROOTSYS/include -L$ROOTSYS/lib `root-config --libs` -lROOTNTuple -fsanitize=address`
```

and run it on the downloaded test file:

```bash
curl -O http://alimonitor.cern.ch/train-workdir/testdata/LFN/alice/data/2015/LHC15o/000246392/pass2/PWGZZ/Run3_Conversion/320_20220701-0813/AOD/174/AO2D.root
./a.out AO2D.root
```

### ROOT version

```
   ------------------------------------------------------------------
  | Welcome to ROOT 6.28/04                        https://root.cern |
  | (c) 1995-2022, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for macosxarm64 on Nov 01 2023, 10:33:00                   |
  | From tags/v6-28-04-alice3@v6-28-04-3-g186ec88ba4                 |
  | With Apple clang version 15.0.0 (clang-1500.0.40.1)              |
  | Try '.help'/'.?', '.demo', '.license', '.credits', '.quit'/'.q'  |
   ------------------------------------------------------------------
```

### Installation method

aliBuild

### Operating system

macos

**Comment by ktf** [comment]
Ok, I think I understand. The tutorial does not handle neither arrays nor VLAs. Arrays support I managed to add myself, is there any example for converting VLAs?

**Comment by jblomer** [comment]
The tutorial on manual data conversion from TTree to RNTuple has been removed now that the importer is available. The importer comes with its own tutorial (`ntpl008_import`). The importer handles VLA correctly. Closing this. Please feel free to reopen if necessary.

**Comment by ktf** [comment]
@jblomer can we get rid of the nagging by the bot, please?
